File: frontend/constants/documento.ts
export const onlyDigits = (value: string): string => String(value ?? '').replace(/\D/g, '');

const allDigitsEqual = (digits: string): boolean => /^(\d)\1+$/.test(digits);

export function isValidCPF(value: string): boolean {
  const cpf = onlyDigits(value);
  if (cpf.length !== 11 || allDigitsEqual(cpf)) return false;

  const calc = (base: string, factor: number): number => {
    let total = 0;
    for (const ch of base) {
      total += Number(ch) * factor--;
    }
    const mod = total % 11;
    return mod < 2 ? 0 : 11 - mod;
  };

  const d1 = calc(cpf.slice(0, 9), 10);
  const d2 = calc(cpf.slice(0, 10), 11);
  return cpf === `${cpf.slice(0, 9)}${d1}${d2}`;
}

export function isValidCNPJ(value: string): boolean {
  const cnpj = onlyDigits(value);
  if (cnpj.length !== 14 || allDigitsEqual(cnpj)) return false;

  const calc = (base: string, weights: number[]): number => {
    const total = base.split('').reduce((sum, ch, idx) => sum + Number(ch) * weights[idx], 0);
    const mod = total % 11;
    return mod < 2 ? 0 : 11 - mod;
  };

  const w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  const w2 = [6, ...w1];
  const d1 = calc(cnpj.slice(0, 12), w1);
  const d2 = calc(cnpj.slice(0, 12) + d1, w2);
  return cnpj === `${cnpj.slice(0, 12)}${d1}${d2}`;
}

File: backend/internal/validation/documento.go
package validation

import "unicode"

func onlyDigits(s string) string {
	out := make([]rune, 0, len(s))
	for _, r := range s {
		if unicode.IsDigit(r) {
			out = append(out, r)
		}
	}
	return string(out)
}

func allDigitsEqual(s string) bool {
	if len(s) == 0 {
		return false
	}
	first := s[0]
	for i := 1; i < len(s); i++ {
		if s[i] != first {
			return false
		}
	}
	return true
}

func IsValidCPF(raw string) bool {
	cpf := onlyDigits(raw)
	if len(cpf) != 11 || allDigitsEqual(cpf) {
		return false
	}

	calc := func(base string, factor int) int {
		total := 0
		for i := 0; i < len(base); i++ {
			total += int(base[i]-'0') * factor
			factor--
		}
		mod := total % 11
		if mod < 2 {
			return 0
		}
		return 11 - mod
	}

	d1 := calc(cpf[:9], 10)
	d2 := calc(cpf[:10], 11)
	return cpf == cpf[:9]+string(rune('0'+d1))+string(rune('0'+d2))
}

func IsValidCNPJ(raw string) bool {
	cnpj := onlyDigits(raw)
	if len(cnpj) != 14 || allDigitsEqual(cnpj) {
		return false
	}

	calc := func(base string, weights []int) int {
		total := 0
		for i := 0; i < len(base); i++ {
			total += int(base[i]-'0') * weights[i]
		}
		mod := total % 11
		if mod < 2 {
			return 0
		}
		return 11 - mod
	}

	w1 := []int{5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}
	w2 := []int{6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}
	d1 := calc(cnpj[:12], w1)
	base2 := cnpj[:12] + string(rune('0'+d1))
	d2 := calc(base2, w2)
	return cnpj == cnpj[:12]+string(rune('0'+d1))+string(rune('0'+d2))
}

